File: scripting/jvmMain/java/com/dropbear/asset/ModelNative.java
package com.dropbear.asset;

import com.dropbear.EucalyptusCoreLoader;
import com.dropbear.asset.model.*;
import java.util.List;

public class ModelNative {
    static {
        new EucalyptusCoreLoader().ensureLoaded();
    }

    public static native String getLabel(long assetManagerHandle, String modelName);
    public static native List<Mesh> getMeshes(long assetManagerHandle, long modelHandle);
    public static native List<Material> getMaterials(long assetManagerHandle, long modelHandle);
    public static native List<Skin> getSkins(long assetManagerHandle, long modelHandle);
    public static native List<Animation> getAnimations(long assetManagerHandle, long modelHandle);
    public static native List<Node> getNodes(long assetManagerHandle, long modelHandle);
}

File: scripting/nativeMain/kotlin/com/dropbear/asset/Model.native.kt
@file:OptIn(ExperimentalForeignApi::class)

package com.dropbear.asset

import com.dropbear.DropbearEngine
import com.dropbear.asset.model.Animation
import com.dropbear.asset.model.Material
import com.dropbear.asset.model.Mesh
import com.dropbear.asset.model.Node
import com.dropbear.asset.model.Skin
import com.dropbear.ffi.generated.*
import kotlin.String
import kotlinx.cinterop.*

actual fun Model.getLabel(): String = memScoped {
    val assets = DropbearEngine.native.assetHandle ?: return@memScoped ""
    val out = alloc<CPointerVar<ByteVar>>()
    val rc = dropbear_asset_model_get_label(assets, id.toULong(), out.ptr)
    if (rc != 0) "" else out.value?.toKString() ?: ""
}

actual fun Model.getMeshes(): List<Mesh> = memScoped {
    TODO("tbc")
//
//    val result = dropbear_asset_model_get_meshes(
//
//    )
}
actual fun Model.getMaterials(): List<Material> = emptyList()
actual fun Model.getSkins(): List<Skin> = emptyList()
actual fun Model.getAnimations(): List<Animation> = emptyList()
actual fun Model.getNodes(): List<Node> = emptyList()
